buffer/src/main/java/io/netty/buffer/api/BufferHolder.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.buffer.api.internal.ResourceSupport;
+import io.netty.buffer.api.internal.Statics;
+
+import java.lang.invoke.VarHandle;
+import java.util.Objects;
+
+import static java.lang.invoke.MethodHandles.lookup;
+
+/**
+ * The {@link BufferHolder} is an abstract class that simplifies the implementation of objects that themselves contain
+ * a {@link Buffer} instance.
+ * <p>
+ * The {@link BufferHolder} can only hold on to a single buffer, so objects and classes that need to hold on to multiple
+ * buffers will have to do their implementation from scratch, though they can use the code of the {@link BufferHolder}
+ * as inspiration. Alternatively, multiple buffers can be
+ * {@linkplain CompositeBuffer#compose(BufferAllocator, Send[]) composed} into a single buffer, which can then be put
+ * in a buffer holder.
+ * <p>
+ * If you just want an object that is a reference to a buffer, then the {@link BufferRef} can be used for that purpose.
+ * If you have an advanced use case where you wish to implement {@link Resource}, and tightly control lifetimes, then
+ * {@link ResourceSupport} can be of help.
+ *
+ * @param <T> The concrete {@link BufferHolder} type.
+ */
+public abstract class BufferHolder<T extends BufferHolder<T>> implements Resource<T> {
+    private static final VarHandle BUF = Statics.findVarHandle(lookup(), BufferHolder.class, "buf", Buffer.class);
```

Comment:
I'm surprised this compiles, Aren't Varhandles JDK 9+ ?

Reply:
Netty 5 is on Java 11.

Reply:
Hmm, I assumed the Netty5 branch is master, README for it still says 8.